mm2src/coins/utxo/rpc_clients/electrum_rpc/connection.rs

```diff
@@ -0,0 +1,872 @@
+            },
+        };
+
+        // Try to connect to the server.
+        let now = Instant::now();
+        let Ok(connect_f) = connect_f.boxed().timeout_secs(timeout).await else {
+            disconnect_and_return!(
+                ElectrumConnectionErr::Timeout(timeout),
+                connection,
+                event_handlers
+            );
+        };
+        // We will use the remaining timeout for version checking.
+        let timeout = (timeout - now.elapsed().as_secs_f64()).max(0.0);
+
+        let stream = disconnect_and_return_if_err!(connect_f, Temporary, connection, event_handlers);
+        disconnect_and_return_if_err!(stream.as_ref().set_nodelay(true), Temporary, connection, event_handlers);
+
+        match secure_connection {
+            true => info!("Electrum client connected to {address} securely"),
+            false => info!("Electrum client connected to {address}"),
+        };
+
+        let (connection_ready_signal, wait_for_connection_ready) = async_oneshot::channel();
+        let connection_loop = {
+            // Branch 1: Disconnect after not receiving responses for too long.
+            let last_response = Arc::new(AtomicU64::new(now_ms()));
+            let no_connection_timeout_f = {
+                let last_response = last_response.clone();
+                async move {
+                    loop {
+                        Timer::sleep(CUTOFF_TIMEOUT).await;
+                        let last_sec = (last_response.load(AtomicOrdering::Relaxed) / 1000) as f64;
+                        if now_float() - last_sec > CUTOFF_TIMEOUT {
+                            break ElectrumConnectionErr::Temporary(format!(
+                                "Server didn't respond for too long ({}s).",
+                                now_float() - last_sec
+                            ));
+                        }
+                    }
+                }
+            };
+            let mut no_connection_timeout_f = Box::pin(no_connection_timeout_f).fuse();
+
+            // Branch 2: Read incoming responses from the server.
+            let (read, mut write) = tokio::io::split(stream);
+            let recv_f = {
+                let connection = connection.clone();
+                let event_handlers = event_handlers.clone();
+                async move {
+                    let mut buffer = String::with_capacity(1024);
+                    let mut buf_reader = BufReader::new(read);
+                    loop {
+                        match buf_reader.read_line(&mut buffer).await {
+                            Ok(c) => {
+                                // FIXME: Why do we close the connection on EOF? Is that a signal from the server
+                                // that no more data will be sent over this connection?
+                                if c == 0 {
+                                    break ElectrumConnectionErr::Temporary("EOF".to_string());
+                                }
```

Comment:
Yes, I think this is a signal from the server, we expected data but got nothing meaning connection is closed, sometimes it happens with incompatible electrums afaik, we got it before from blockstream electrums here https://1209k.com/bitcoin-eye/ele.php?chain=btc. Why did we change it from this on dev btw https://github.com/KomodoPlatform/komodo-defi-framework/blob/df6ab98b3b2ee84918bfc0bf1bc03d1e1f219985/mm2src/coins/utxo/rpc_clients.rs#L2850-L2853 where we show the address of the server disconnecting.

Reply:
all these errors without addresses are catched/returned and formatted and logged with the address.

https://github.com/KomodoPlatform/komodo-defi-framework/blob/76b2f689066a83d8d2d8838c2f33770eed03b16a/mm2src/coins/utxo/rpc_clients/electrum_rpc/connection.rs#L537-L543